trustfall_core/src/graphql_query/query.rs

```diff
@@ -125,49 +125,52 @@ fn try_get_query_root(document: &ExecutableDocument) -> Result<&Positioned<Field
 
     match &document.operations {
         DocumentOperations::Multiple(mult) => {
-            return Err(ParseError::MultipleOperationsInDocument(mult.values().next().unwrap().pos))
+            return if mult.values().len() > 1 {
+                Err(ParseError::MultipleOperationsInDocument(mult.values().next().unwrap().pos))
+            } else {
+                let node = mult.values().next().unwrap();
+                parse_operation_definition(node)
+            }
         }
-        DocumentOperations::Single(op) => {
-            let root_node = &op.node;
+        DocumentOperations::Single(op) => parse_operation_definition(op),
+    }
+}
 
-            if root_node.ty != OperationType::Query {
-                return Err(ParseError::DocumentNotAQuery(op.pos));
-            }
+fn parse_operation_definition(
+    op: &Positioned<OperationDefinition>,
+) -> Result<&Positioned<Field>, ParseError> {
+    let root_node = &op.node;
 
-            if !root_node.variable_definitions.is_empty() {
-                let first_variable_definition = root_node.variable_definitions.first().unwrap();
-                return Err(ParseError::OtherError(
-                    "Found GraphQL query variable definitions. \
-                    These are not necessary since variables are defined implicitly, \
-                    and must be removed."
-                        .to_string(),
-                    first_variable_definition.pos,
-                ));
-            }
-            if !root_node.directives.is_empty() {
-                let first_directive = root_node.directives.first().unwrap();
-                return Err(ParseError::DirectiveNotInsideQueryRoot(
-                    first_directive.node.name.node.to_string(),
-                    first_directive.pos,
-                ));
-            }
+    if root_node.ty != OperationType::Query {
+        return Err(ParseError::DocumentNotAQuery(op.pos));
+    }
 
-            let root_selection_set = &root_node.selection_set.node;
-            let root_items = &root_selection_set.items;
-            if root_items.len() != 1 {
-                return Err(ParseError::MultipleQueryRoots(root_items[1].pos));
-            }
+    if !root_node.variable_definitions.is_empty() {
+        let first_variable_definition = root_node.variable_definitions.first().unwrap();
+        return Err(ParseError::VariableDefinitionInRootQuery(first_variable_definition.pos));
+    }
+    if !root_node.directives.is_empty() {
```

Comment:
These should both probably be `if let Some()` as well. Again, this is some of the first Rust I ever wrote, and I'm kind of embarrassed by it at this point 😅

Reply:
@obi1kenobi these are actually of type `Vec` so I think `.is_empty()` should be fine here. 

<img width="526" alt="Screenshot 2024-04-20 at 9 33 52 AM" src="https://github.com/obi1kenobi/trustfall/assets/22461703/c7a76e25-9d85-4e85-88e6-b7ac3e00447e">

Reply:
Ah, sorry, I should have been clearer. We're doing an `.is_empty()` followed by `.first().unwrap()`.

This is the more elegant, panic-free way to do the same:
```rust
if let Some(first_variable_definition) = root_node.variable_definitions.first() {
    return Err(ParseError::VariableDefinitionInRootQuery(first_variable_definition.pos));
}
```

I just don't think I knew about `if let` at the time I wrote this code.